// wdio-selenium-standalone-service/build/launcher.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _seleniumStandalone = require('selenium-standalone');

var _seleniumStandalone2 = _interopRequireDefault(_seleniumStandalone);

var _mkdirp = require('mkdirp');

var _mkdirp2 = _interopRequireDefault(_mkdirp);

var _path = require('path');

var _path2 = _interopRequireDefault(_path);

var _fs = require('fs');

var _fs2 = _interopRequireDefault(_fs);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

var DEFAULT_LOG_FILENAME = 'selenium-standalone.txt';

var SeleniumStandaloneLauncher = function () {
    function SeleniumStandaloneLauncher() {
        _classCallCheck(this, SeleniumStandaloneLauncher);

        this.seleniumLogs = null;
        this.logToStdout = false;
        this.opts = {};
    }

    _createClass(SeleniumStandaloneLauncher, [{
        key: 'onPrepare',
        value: function onPrepare(config) {
            var _this = this;

            this.opts.seleniumArgs = config.seleniumArgs || [];
            this.seleniumLogs = config.seleniumLogs;
            this.logToStdout = config.logToStdout;

            return new Promise(function (resolve, reject) {
                return _seleniumStandalone2.default.start(_this.opts, function (err, process) {
                    if (err) {
                        return reject(err);
                    }

                    _this.process = process;
                    if (typeof _this.seleniumLogs === 'string') {
                        _this._redirectLogStream();
                    }

                    resolve();
                });
            });
        }
    }, {
        key: 'onComplete',
        value: function onComplete() {
            if (this.process) {
                this.process.kill();
            }
        }
    }, {
        key: '_redirectLogStream',
        value: function _redirectLogStream() {
            var logFile = _path2.default.join(process.cwd(), this.seleniumLogs);

            /**
             * create directory if not existing
             */
            try {
                _fs2.default.statSync(logFile);
            } catch (e) {
                _mkdirp2.default.sync(logFile);
            }

            if (logFile.slice(-1) === '/') {
                logFile += DEFAULT_LOG_FILENAME;
            }

            var logStream = _fs2.default.createWriteStream(logFile, { flags: 'w' });
            this.process.stdout.pipe(logStream);
            this.process.stderr.pipe(logStream);
        }
    }]);

    return SeleniumStandaloneLauncher;
}();

exports.default = SeleniumStandaloneLauncher;
